### PR TITLE
16.0 mig sale delivery state

### PR DESCRIPTION
**Pull request by sebastienbeau**
Following the discussion here : https://github.com/OCA/sale-workflow/pull/2361

As Odoo have introduced the state "delivery_status" in V16, we believed that the module was useless.
But as the native implementation is broken so we still need this module.
see comment here : https://github.com/OCA/sale-workflow/pull/2361#issuecomment-1801656951

In order to not add any extra new fields, the feature is split in 3 modules
- sale_delivery_state (this add the field but it's not visible in the view)
- sale_stock_delivery_state (glue module to overwrite the implementation of the computed field delivery_status)
- sale_delivery_state_view (glue module when sale_stock is not installed, it will show the field)

It's sad to have to do 3 modules for this, but we do not have choice...

### CODE REVIEW

**Comment by sebastienbeau** [comment]
@rousseldenis @lmignon needed for having clean state on shopinvader

**Comment by flotho** [comment]
hi @sebastienbeau , do you need a review here ?
Regards

File: sale_stock_delivery_state/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    "license": "AGPL-3",
+    "external_dependencies": {
+        "python": [],
+        "bin": [],
+    },
+    "depends": [
+        "sale_stock",
+        "sale_delivery_state",
+    ],
+    "data": [],
```
Comment (yostashiro) [inline sale_stock_delivery_state/__manifest__.py]:
```suggestion
    "auto_install": True,
    "data": [],
```

**Review by yostashiro — COMMENTED** [review]
How about integrating `sale_delivery_state_view` into `sale_delivery_state`, and make the added field invisible in `sale_stock_delivery_state`?

**Comment by sebastienbeau** [comment]
@yostashiro seem to be  good idea, I going to change it today
@flotho a review will be welcome (maybe just after the change propose by @yostashiro)

**Comment by sebastienbeau** [comment]
@flotho @yostashiro the PR is ready for the review. I will squash my commit before merging it

**Review by lmignon — APPROVED** [review]
Code review only

**Comment by sebastienbeau** [comment]
@flotho @yostashiro ping, thanks in advance for your feedback

**Comment by sebastienbeau** [comment]
@Kev-Roche can you review?

File: sale_delivery_state/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+is set, it will trigger the computation of delivery state.
+
+This module also works with delivery.carrier fees that are added as a
+sale order line. Thoses line are special as they will never be considered delivered.
+Delivery fees lines are ignored in the computation of the delivery state.
+
+
+This module do not add the "views" for the field "delivery_status".
+If you have the sale_stock module installed please installed "sale_stock_delivery_state".
+If not install the module "sale_delivery_state_view".
```
Comment (yostashiro) [inline sale_delivery_state/readme/DESCRIPTION.rst]:
```suggestion
When the 'sale_stock' module is installed, the glue module 'sale_stock_delivery_state' should also be installed; this module is designed to override the compute method of the delivery status field from 'sale_stock'.
```

**Review by yostashiro — COMMENTED** [review]
@sebastienbeau Would you consider adding a migration script?

**Comment by sebastienbeau** [comment]
@yostashiro I have added a post-init-hook. This should solve
- migration from 14 to 16 => the field is recomputed so we have the right value
- new installation on v16 => after the installation on an db that have sale_stock, the field is recomputed so all existing Sale Order will have the right value.

It's ok for you ?

**Comment by yostashiro** [comment]
@sebastienbeau Thanks for the update.  I think `sale_delivery_state` still requires a migration script for the scenario of upgrading a database with this module but no `sale_stock`.  Simply moving `post_init_hook()` from `sale_stock_delivery_state` to `sale_delivery_state` might do the job, but I wasn't perfectly sure.

**Comment by sebastienbeau** [comment]
@yostashiro When migrating a database to V16, as the field "delivery_status" have changed of name ("delivery_state" before),
Odoo will consider it as a new field and will compute it and fill the value on all existing sale order.
So everything should work without any extra logic.

**Comment by yostashiro** [comment]
> @yostashiro When migrating a database to V16, as the field "delivery_status" have changed of name ("delivery_state" before), Odoo will consider it as a new field and will compute it and fill the value on all existing sale order. So everything should work without any extra logic.

@sebastienbeau Ah, I see. Missed that point. Thanks for the explanation!

**Review by yostashiro — APPROVED** [review]

**Comment by sebastienbeau** [comment]
Thanks for the review let's merge it

/ocabot merge path

**Comment by ValentinVinagre** [comment]
Hey @pedrobaeza @sebastienbeau @yostashiro @HaraldPanten @etobella 
Hello,
I thought that Oca was going to go down the path of using Odoo functionality. Will the native Odoo functionality be left aside in the end? According to the conversation with Pedrobaeza, it seemed that the way forward was going to be to use the native way of Odoo.

The discusion: https://github.com/OCA/purchase-workflow/pull/2081

**Comment by pedrobaeza** [comment]
I prefer Odoo native feature, but there's a lot of people in OCA, and they can think different. At least a mention in the README should be done about the differences between this module and the standard. Or maybe they didn't know about the core feature?

**Comment by HaraldPanten** [comment]
I think they knew about the native Odoo feature, but the decision was to keep the OCA one, because it's more complete than de native one.

**Pull request merged**